client_proxy_server_test.go

```diff
@@ -623,51 +623,54 @@ func (ts *testServer) Close() {
 	}
 }
 
-// websocketEchoHandler upgrades the connection associated with the request, and
-// echoes binary messages read off the websocket connection back to the client.
-var websocketEchoHandler = http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-	upgrader := Upgrader{
-		CheckOrigin: func(r *http.Request) bool {
-			return true // Accepting all requests
-		},
-		Subprotocols: []string{
-			subprotocolV1,
-			subprotocolV2,
-		},
-	}
-	wsConn, err := upgrader.Upgrade(w, req, nil)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-	}
-	defer wsConn.Close()
-	for {
-		writer, err := wsConn.NextWriter(BinaryMessage)
-		if err != nil {
-			break
+// newWebsocketEchoHandler returns a handler that upgrades the connection associated with the request,
+// and echoes binary messages read off the websocket connection back to the client.
+func newWebsocketEchoHandler(t *testing.T) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		upgrader := Upgrader{
+			CheckOrigin: func(r *http.Request) bool {
+				return true // Accepting all requests
+			},
+			Subprotocols: []string{
+				subprotocolV1,
+				subprotocolV2,
+			},
 		}
-		messageType, reader, err := wsConn.NextReader()
+		wsConn, err := upgrader.Upgrade(w, req, nil)
 		if err != nil {
-			break
-		}
-		if messageType != BinaryMessage {
-			http.Error(w, "websocket reader not binary message type",
-				http.StatusInternalServerError)
+			t.Logf("websocketEchoHandler Upgrade: %v, %#v", err, req)
+			return
```

Comment:
this fixed a spurious panic in the test where Upgrade failed (due to the missing buffer handling below) and wsConn was nil

also, we can't call http.Error after upgrade.Upgrade since headers are already sent

Reply:
yep https://pkg.go.dev/net/http#Error